### PR TITLE
[NSTK-1085]PSO QoS Support for FA

### PR DESCRIPTION
**Pull request by david-enli**
QoS is introduced for PSO 6.0.0+, specifically we enable user to inpose bandwidth_limit and iops_limit on storageclass level. 

*FA only
*Purity version 5.3.0+ only

### CODE REVIEW

File: docs/csi-qos-spec.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
+# Apply QoS control on storageclass level
+
+## Use Cases
+- Users want to set bandwidth limit for their volumes
+- Users want to set IOPS limit for their volumes
+
+## Dependencies
+* Pure CSI driver 6.0.0+ 
+* Purity Version 5.3.0+ (Public API version 1.17 +)
```
Comment (sdodsley) [inline docs/csi-qos-spec.md:9]: Call that REST API, not Public API

File: docs/csi-qos-spec.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+* Support FlashArray block storage only
```
Comment (sdodsley) [inline docs/csi-qos-spec.md:10]: Support **for** FlashArray block storage only.
Like you said block storage just to cover for FA-files in the future

File: docs/csi-qos-spec.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
+# Apply QoS control on storageclass level
```
Comment (sdodsley) [inline docs/csi-qos-spec.md:1]: Call it Storage Quality-of-Service (QoS) and then you can use the acronym throughout the rest of the document

File: docs/csi-qos-spec.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+* Pure CSI driver 6.0.0+ 
+* Purity Version 5.3.0+ (Public API version 1.17 +)
+* Support FlashArray block storage only
+
+## General guidance
+QoS allows users to impose QoS limits on their persistent volumes created and managed by PSO on Kubernetes clusters. Whenever throughput exceeds the limit specified, throttling occurs. Specifically, we enable users to specify their bandwidth limit in terms of volume capacity per second and Input/Output per second. Users can opt to enable one or the other, or both. The two parameters are passed in through parameter fields in storageclasses. The feature is only supported on FlashArray block storage. 
+
+At volume provision time:
+1. If the user does not own any FlashArray that supports QoS (5.3 and above), the volume will NOT be provisioned successfully. 
+2. If the user owns multiple devices, PSO will look for the FlashArray that supports QoS and provision the volume on that device. If multiple arrays support QoS, the existing algorithm is leveraged to determine the array to provision the volume.
```
Comment (sdodsley) [inline docs/csi-qos-spec.md:17]: How do the storage topology flags implemented - is the QoS API version check made before or after the topology checks?

File: docs/csi-qos-spec.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+    add `bandwidth_limit` as a parameter in storageclass
+
+4. To enable IOPS limit
+
+    add `iops` as a parameter in storageclass
+
+## Restrictions
+* Both parameters must be passed in as string types, i.e., they need to have double quotation marks around them, see examples in the later section.
+* If the bandwidth limit is set, it must be between 1 MB/s and 512 GB/s. Enter the size as a number (bytes) or number with a single character unit symbol. Valid unit symbols are K, M, G, T, P, representing KiB, MiB, GiB, TiB, and PiB, respectively, where "Ki" denotes 2^10, "Mi" denotes 2^20, and so on. If the unit symbol is not specified, the unit defaults to bytes. And when unit symbol is not specified, the number entered must be multiple of 512.
```
Comment (sdodsley) [inline docs/csi-qos-spec.md:28]: I suspect you need to be a bit more clear here.
If the max is 512GB/s then why say that TiB and PiB are valid strings?
If no minimum is provided then the lowest value that can be entered is 1048576. your description sort of implies 512, 1024, etc are acceptable values.

File: docs/csi-qos-spec.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+3. To enable bandwidth limit
+
+    add `bandwidth_limit` as a parameter in storageclass
+
+4. To enable IOPS limit
+
+    add `iops` as a parameter in storageclass
+
+## Restrictions
+* Both parameters must be passed in as string types, i.e., they need to have double quotation marks around them, see examples in the later section.
```
Comment (sdodsley) [inline docs/csi-qos-spec.md:27]: why does iops have to be a string - this is a number

File: docs/csi-qos-spec.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+* If the bandwidth limit is set, it must be between 1 MB/s and 512 GB/s. Enter the size as a number (bytes) or number with a single character unit symbol. Valid unit symbols are K, M, G, T, P, representing KiB, MiB, GiB, TiB, and PiB, respectively, where "Ki" denotes 2^10, "Mi" denotes 2^20, and so on. If the unit symbol is not specified, the unit defaults to bytes. And when unit symbol is not specified, the number entered must be multiple of 512. 
+
+
+ 
+## Example
+1. **Verify that user devices support QoS**
```
Comment (sdodsley) [inline docs/csi-qos-spec.md:33]: I hope we are doing this in PSO as part of our validation. I don't think should expect the customer to do this unless it is part of a troubleshooting process.

File: docs/csi-qos-spec.md [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+2. If the user owns multiple devices, PSO will look for the FlashArray that supports QoS and provision the volume on that device. If multiple arrays support QoS, the existing algorithm is leveraged to determine the array to provision the volume. 
+3. To enable bandwidth limit
+
+    add `bandwidth_limit` as a parameter in storageclass
+
+4. To enable IOPS limit
+
+    add `iops` as a parameter in storageclass
+
+## Restrictions
```
Comment (sdodsley) [inline docs/csi-qos-spec.md:26]: Add a restriction stating the limits for IOPs value are 100 to 100000000

File: docs/csi-qos-spec.md [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    ```yaml
+    kind: StorageClass
+    apiVersion: storage.k8s.io/v1
+    metadata:
+      name: bronze
+      labels:
+        kubernetes.io/cluster-service: "true"
+    provisioner: pure-csi # This must match the name of the CSIDriver. And the name of the CSI plugin from the RPC 'GetPluginInfo'
+    parameters:
+      #TODO: choose limit
```
Comment (sdodsley) [inline docs/csi-qos-spec.md:71]: need this line?

File: docs/csi-qos-spec.md [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    ```
+
+    ```yaml
+    kind: StorageClass
+    apiVersion: storage.k8s.io/v1
+    metadata:
+      name: bronze
+      labels:
+        kubernetes.io/cluster-service: "true"
+    provisioner: pure-csi # This must match the name of the CSIDriver. And the name of the CSI plugin from the RPC 'GetPluginInfo'
```
Comment (sdodsley) [inline docs/csi-qos-spec.md:69]: Not sure the comment is needed here - the last sentence is a bit technical

File: docs/csi-qos-spec.md [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+      # Referenced in pod.yaml for the volume spec
+      # TODO: choose name for PVC
+      name: testingPVC
+    spec:
+      accessModes:
+        - ReadWriteOnce
+      resources:
+        requests:
+          storage: 2Gi
+      # TODO: Matches the name defined in deployment/storageclass.yaml
```
Comment (sdodsley) [inline docs/csi-qos-spec.md:97]: TODO comment is duplicating the comment on the next line

File: docs/csi-qos-spec.md [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+    To apply:
+        ```
+        kubectl apply -f {PVC_yaml_file_name}
+        ```. At this point, QoS is not a built-in support for CSI drivers from Kubernetes, so it is entirely being managed by PSO, therefore, to see if the QoS is set successfully, try the following steps:
+    
+    First run ```kubectl get pvc``` to make sure your volume is bounded and grab the volume name. QoS setting can be tracked in two ways
+    1. through the command line
+            ```
+            GET https://{deviceIP}}/api/{api_version}}/volume/{volume_name}}?qos=true
```
Comment (sdodsley) [inline docs/csi-qos-spec.md:109]: This isn't a command-line call - this is a REST API call - you need to provide a true Linux command

File: docs/csi-qos-spec.md [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+            ```
+    2. through the dashboard UI. 
+    go to Storage -> Volumes, search for the volume just created and QoS is shown at the bottom. 
+
+**Note:** this feature applies QoS limits at volume-creation time, but both bandwidth and IOPS limit can be modified at any time after volume-creation either through the dashboard UI or REST API ```Put Volume``` command, see API guide for details.
```
Comment (sdodsley) [inline docs/csi-qos-spec.md:114]: Just for giggles, reference the Pure Storage FlashArray Ansible module `purefa_volume` to affect QoS changes.

File: docs/csi-qos-spec.md (hunk 1 above)
Comment (david-enli) [reply docs/csi-qos-spec.md:9]: Will do

File: docs/csi-qos-spec.md (hunk 2 above)
Comment (david-enli) [reply docs/csi-qos-spec.md:10]: will do

File: docs/csi-qos-spec.md (hunk 4 above)
Comment (david-enli) [reply docs/csi-qos-spec.md:17]: The code change is not complete yet, let me check the code and get back to you

File: docs/csi-qos-spec.md (hunk 7 above)
Comment (david-enli) [reply docs/csi-qos-spec.md:33]: yes, this is done in the code, this is just here for information, I will mention that

File: docs/csi-qos-spec.md (hunk 4 above)
Comment (david-enli) [reply docs/csi-qos-spec.md:17]: API version check is after the topology logic. Topology check is in k8s, and API version check is in NSM, basicaly it rediscovers all backends that have API version 1.17 and above and return them as the backend devices to apply our algorithm to find the array to provision

**Comment by david-enli** [comment]
changes are made by @carysyd and @sdodsley on hackmd, consolidated in latest commit.

File: docs/csi-qos-spec.md [hunk 14]
```diff
@@ -0,0 +1,104 @@
+# [PSO Enhancement Proposal] Apply Storage Quality-of-Service (QoS) Control on PSO StorageClass
```
Comment (sdodsley) [inline docs/csi-qos-spec.md:1]: Remove the [] stuff

Comment (david-enli) [reply docs/csi-qos-spec.md]: oops, sry, removed.

**Pull request merged**